Add Nix flake and HLS setup

gabrielhdt opened the pull request:
This PR adds a flake to the repository, and a `hie` file to setup the language server.

The flake is not fully operational because, if I understand correctly, `callCabal2nix` (which is used to transform cabal dependencies into Nix dependencies) fails on `simple-smt-z3` because its tests use `simple-smt-tests`, but the latter is not a library.

gabrielhdt committed:
Add Nix flake and HLS setup

qaristote commented:
> because its tests use simple-smt-tests, but the latter is not a library

[It is a library](https://github.com/tweag/simple-smt/blob/master/tests/simple-smt-tests.cabal), what do you mean by that ?

qaristote commented:
Also building `simple-smt-z3` will require `pkgs.z3.dev`. [In Pirouette](https://github.com/tweag/pirouette/blob/69d1eea06fcafd58244f129e564c8d34825bc35a/nix/packages.nix#L32) it is made available as a `nativeBuildInput`.

gabrielhdt commented:
> > because its tests use simple-smt-tests, but the latter is not a library
> 
> [It is a library](https://github.com/tweag/simple-smt/blob/master/tests/simple-smt-tests.cabal), what do you mean by that ?

All right I didn't see that, my apologies. I'll investigate some more then.

gabrielhdt commented:
I can't figure out why, but the dependency on `simple-smt-tests` disturbs `callCabal2nix` which fails with
```
error: anonymous function at /nix/store/swsnfnd2lqsg0199m65yb2nwk333jarm-cabal2nix-simple-smt-z3/default.nix:1:1 called without required argument 'simple-smt-tests'

       at /nix/store/pvxhlilwj6bmrsqq838zw29agd8rjksf-source/pkgs/development/haskell-modules/make-package-set.nix:97:40:

           96|       # this wraps the `drv` function to add a `overrideScope` function to the result.
           97|       drvScope = allArgs: ensureAttrs (drv allArgs) // {
             |                                        ^
           98|         overrideScope = f:
```

gabrielhdt committed:
Add simple-smt-tests and simple-smt-z3

callCabal2nix fails on simple-smt-z3 because of the
dependency on simple-smt-tests (in the test suite)

qaristote commented:
> I can't figure out why, but the dependency on simple-smt-tests disturbs callCabal2nix which fails

I've never used `haskell.nix`, but my guess is that `callCabal2nix` tries to resolve dependencies using `nixpkgs`, which itself inherits its Haskell packages from Hackage. Since `simple-smt-tests` is only a local package for now, you probably need to create some sort of overlay to populate `nixpkgs` with `simple-smt-tests`.

qaristote commented:
Maybe `simple-smt-z3 = hpkgs.callCabal2nix "simple-smt-z3" ./Z3 { inherit simple-smt-tests; };` is enough ?

gabrielhdt committed:
Fix simple-smt-z3

gabrielhdt committed:
Add shellhook for libz3
